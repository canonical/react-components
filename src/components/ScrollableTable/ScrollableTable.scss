@import "vanilla-framework/scss/settings";

/* stylelint-disable selector-max-type */
.scrollable-table {
  table {
    margin-bottom: 0;
  }

  th,
  td {
    display: table-cell;
    vertical-align: top;
  }

  th::before {
    content: "";
    height: 1rem;
  }

  thead,
  tbody {
    display: block;
    overflow: hidden auto;
    scrollbar-gutter: stable;
  }

  tbody {
    height: calc(100dvh - 323px);
  }

  thead tr,
  tbody tr {
    display: table;
    table-layout: fixed;
    width: 100%;
  }

  @media screen and (width < $breakpoint-large) {
    .p-table--mobile-card {
      thead {
        display: none;
      }

      td {
        display: block;
      }

      tbody tr {
        display: block;
      }

      tbody tr:first-child {
        margin-top: $spv--x-large;
      }
    }
  }
}
/* stylelint-enable selector-max-type */
