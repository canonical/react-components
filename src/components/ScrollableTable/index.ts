export { default } from "./ScrollableTable";
export type { Props as ScrollableTableProps } from "./ScrollableTable";
