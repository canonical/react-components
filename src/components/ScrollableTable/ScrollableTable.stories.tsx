import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import ScrollableTable from "./ScrollableTable";
import { MainTable } from "../../index";
import { Description, Subtitle, Title } from "@storybook/blocks";

const meta: Meta<typeof ScrollableTable> = {
  component: ScrollableTable,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof ScrollableTable>;

const StoryExample = (args: Story["args"]) => {
  const headers = [
    { content: "Fact" },
    { content: "Relevancy score" },
    { content: "Size" },
    { content: "ID" },
    { "aria-label": "Actions", className: "actions" },
  ];

  const facts = [
    "Dragons are mythical creatures",
    "They can fly",
    "They breathe fire",
    "They hoard treasure",
    "They are often depicted as large reptiles",
    "They appear in various cultures' folklore",
    "They can be friendly or hostile",
    "They are often associated with wisdom",
    "They can shapeshift in some legends",
    "They are sometimes guardians of sacred places",
    "They can be found in literature and movies",
    "They are often depicted with wings",
    "They can be of various colors",
    "They are sometimes associated with knights",
  ];
  const rows = facts.map((item) => {
    return {
      columns: [
        { content: item, role: "rowheader" },
        { content: 1 },
        { content: "1 GiB" },
        { content: 2 },
      ],
    };
  });

  return (
    <div>
      <h1>Above contents</h1>
      <ScrollableTable
        belowIds={args.belowIds}
        dependencies={[headers, rows]}
        tableId="facts-about-dragons"
      >
        <MainTable headers={headers} rows={rows} id="facts-about-dragons" />
      </ScrollableTable>
      <div id="footer">
        <h2 className="u-no-margin">Below contents</h2>
        <div>here be dragons.</div>
      </div>
    </div>
  );
};

export const Default: Story = {
  args: {
    belowIds: ["footer"],
  },
  render: StoryExample,
  parameters: {
    docs: {
      page: () => (
        <>
          <Title />
          <Subtitle />
          <Description />
        </>
      ),
    },
  },
};
