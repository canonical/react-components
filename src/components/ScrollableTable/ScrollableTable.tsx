import React from "react";
import type { DependencyList, FC, ReactNode } from "react";
import { useEffect } from "react";
import { getAbsoluteHeightBelowById, getParentsBottomSpacing } from "utils";
import "./ScrollableTable.scss";
import { useListener } from "hooks";

export type Props = {
  /**
   * A MainTable that will be made scrollable.
   */
  children: ReactNode;

  /**
   * An array of dependencies that will trigger a re-calculation of the table body height when they change.
   */
  dependencies: DependencyList;

  /**
   * The ID of the table element that contains the `<tbody>` to be made scrollable.
   */
  tableId: string;

  /**
   * An array of element IDs below the table that should be considered when calculating the height.
   */
  belowIds?: string[];
};

/**
 * This is a [React](https://reactjs.org/) component for use within the [Vanilla framework](https://docs.vanillaframework.io/).
 *
 * It is used to make a table scrollable by adjusting the height of the `<tbody>` element based on the viewport height and the heights of elements above and below it. As a result the header is sticky and only the body scrolls
 */
const ScrollableTable: FC<Props> = ({
  dependencies,
  children,
  tableId,
  belowIds = [],
}) => {
  const updateTBodyHeight = () => {
    const table = document.getElementById(tableId);
    if (!table || table.children.length !== 2) {
      return;
    }

    const tBody = table.children[1];
    const above = tBody.getBoundingClientRect().top + 1;
    const below = belowIds.reduce(
      (acc, belowId) => acc + getAbsoluteHeightBelowById(belowId),
      0,
    );
    const parentsBottomSpacing = getParentsBottomSpacing(table);
    const offset = Math.ceil(above + below + parentsBottomSpacing);
    const style = `height: calc(100dvh - ${offset}px); min-height: calc(100dvh - ${offset}px)`;
    tBody.setAttribute("style", style);
  };

  useListener(window, updateTBodyHeight, "resize", true);
  useEffect(updateTBodyHeight, [...dependencies, belowIds, tableId]);

  return <div className="scrollable-table">{children}</div>;
};

export default ScrollableTable;
