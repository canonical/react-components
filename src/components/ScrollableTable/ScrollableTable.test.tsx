import React from "react";
import { render, screen } from "@testing-library/react";

import ScrollableTable from "./ScrollableTable";
import { MainTable } from "../../index";

const headers = [{ content: "Fact" }];
const rows = [
  {
    columns: [{ content: "foo" }],
  },
];

it("renders", () => {
  render(
    <ScrollableTable dependencies={[]} tableId="test-table">
      <MainTable headers={headers} rows={rows} id="test-table" />
    </ScrollableTable>,
  );
  expect(screen.getByText("Fact")).toBeInTheDocument();
});

it("respects above spacing", () => {
  render(
    <div role="main">
      <h1>Above</h1>
      <ScrollableTable dependencies={[]} tableId="test-table">
        <MainTable headers={headers} rows={rows} id="test-table" />
      </ScrollableTable>
    </div>,
  );
  expect(screen.getByRole("main")).toMatchSnapshot();
});

it("respects below spacing", () => {
  render(
    <div role="main">
      <h1>Above</h1>
      <ScrollableTable
        dependencies={[]}
        belowIds={["footer"]}
        tableId="test-table"
      >
        <MainTable headers={headers} rows={rows} id="test-table" />
      </ScrollableTable>
      <div id="footer">
        <h2>Below</h2>
      </div>
    </div>,
  );
  expect(screen.getByRole("main")).toMatchSnapshot();
});
