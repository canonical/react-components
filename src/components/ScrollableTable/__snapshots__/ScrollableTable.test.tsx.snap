// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`respects above spacing 1`] = `
<div
  role="main"
>
  <h1>
    Above
  </h1>
  <div
    class="scrollable-table"
  >
    <table
      class=""
      id="test-table"
      role="grid"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            role="columnheader"
          >
            Fact
          </th>
        </tr>
      </thead>
      <tbody
        style="height: calc(100dvh - 9px); min-height: calc(100dvh - 9px)"
      >
        <tr
          role="row"
        >
          <td
            aria-hidden="false"
            class=""
            role="gridcell"
          >
            foo
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`respects below spacing 1`] = `
<div
  role="main"
>
  <h1>
    Above
  </h1>
  <div
    class="scrollable-table"
  >
    <table
      class=""
      id="test-table"
      role="grid"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            role="columnheader"
          >
            Fact
          </th>
        </tr>
      </thead>
      <tbody
        style="height: calc(100dvh - 10px); min-height: calc(100dvh - 10px)"
      >
        <tr
          role="row"
        >
          <td
            aria-hidden="false"
            class=""
            role="gridcell"
          >
            foo
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    id="footer"
  >
    <h2>
      Below
    </h2>
  </div>
</div>
`;
